modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxRecoveryWithConcurrentRollbackTest.java

```diff
@@ -258,6 +264,129 @@ else if (g1Keys.contains(key))
         assertEquals(s1, s2);
     }
 
+
+    /**
+     * The test enforces the concurrent processing of the same prepared transaction both in the
+     * tx recovery procedure started due to primary node left and in the tx recovery request handler
+     * invoked by message from another backup node.
+     * <p>
+     * The idea is to have a 3-nodes cluster and a cache with 2 backups. So there will be 2 backup nodes
+     * to execute the tx recovery in parallel if primary one would fail. These backup nodes will send the
+     * tx recovery requests to each other, so the tx recovery request handler will be invoked as well.
+     * <p>
+     * Use several attempts to reproduce the race condition.
+     * <p>
+     * Expected result: transaction is finished on both backup nodes and the partition map exchange is completed as well.
+     */
+    @Test
+    public void testRecoveryNotDeadLockOnPrimaryFail() throws Exception {
+        backups = 2;
```

Comment:
Never used?

Reply:
Used in TxRecoveryWithConcurrentRollbackTest::getConfiguration to create node config